firewood/src/merkle/node.rs

```diff
@@ -717,6 +718,176 @@ impl Storable for Node {
+                    )))?,
+                };
+                let data = match items.next().unwrap() {
+                    Encoded::Raw(data) => data,
+                    Encoded::Data(data) => Err(D::Error::custom(format!(
+                        "incorrect encoded type for leaf node data {:?}",
+                        data
+                    )))?,
+                };
+                let path_nibbles = Nibbles::<0>::new(&path);
+                let path = PartialPath::from_nibbles(path_nibbles.into_iter()).0;
+                let node = EncodedNodeType::Leaf(LeafNode {
+                    path,
+                    data: Data(data),
+                });
+                Ok(Self {
+                    node,
+                    phantom: PhantomData,
+                })
+            }
+            BRANCH_NODE_SIZE => {
+                let mut items = items;
+
+                // we've already validated the size, that's why we can safely unwrap
+                let data = match items.pop().unwrap() {
+                    Encoded::Raw(data) => Err(D::Error::custom(format!(
+                        "incorrect encoded type for branch node value {:?}",
+                        data
+                    )))?,
+                    Encoded::Data(data) => other
+                        .deserialize(data.as_ref())
+                        .map_err(|e| D::Error::custom(format!("bincode error: {e}")))?,
+                };
+                // Extract the value of the branch node and set to None if it's an empty Vec
+                let value = Some(Data(data)).filter(|data| !data.is_empty());
+
+                let mut children: [Option<Vec<u8>>; MAX_CHILDREN] = Default::default();
+
+                // we popped the last element, so their should only be NBRANCH items left
+                for (i, chd) in items.into_iter().enumerate() {
+                    let chd = match chd {
+                        Encoded::Raw(chd) => chd,
+                        Encoded::Data(chd) => other
+                            .deserialize(chd.as_ref())
+                            .map_err(|e| D::Error::custom(format!("bincode error: {e}")))?,
+                    };
+                    children[i] = Some(chd).filter(|chd: &Vec<u8>| !chd.is_empty());
+                }
+                let node = EncodedNodeType::Branch {
+                    children: children.into(),
+                    value,
+                };
+                Ok(Self {
+                    node,
+                    phantom: PhantomData,
+                })
+            }
+            size => Err(D::Error::custom(format!("invalid size: {size}"))),
+        }
+    }
+}
+pub trait BinarySerde {
+    type SerializeError: Debug;
+    type DeserializeError: Debug;
+
+    fn serialize<T: Serialize>(&self, t: &T) -> Result<Vec<u8>, Self::SerializeError>;
+
+    fn deserialize<'de, T: Deserialize<'de>>(
+        &self,
+        bytes: &'de [u8],
+    ) -> Result<T, Self::DeserializeError>;
+}
```

Comment:
I'm thinking it would be best if there were only one concrete way to create the serializer.
```suggestion
pub trait BinarySerde {
    type SerializeError: serde::ser::Error;
    type DeserializeError: serde::de::Error;
    
    fn new() -> Self;
    
   fn serialize<T: Serialize>(t: &T) -> Result<Vec<u8>, Self::SerializeError> {
       Self::new().serialize_impl(t)
   }

    fn deserialize<'de, T: Deserialize<'de>>(bytes: &'de [u8],) -> Result<T, Self::DeserializeError> {
        Self::new().deserialize_impl(bytes)
    }

    fn serialize_impl<T: Serialize>(t: &T) -> Result<Vec<u8>, Self::SerializeError>;

    fn deserialize_impl<'de, T: Deserialize<'de>>(
        &self,
        bytes: &'de [u8],
    ) -> Result<T, Self::DeserializeError>;
}
```

I'm not sure that this is *the best* solution, but you wan't to make sure there's only one way to construct the serializer.